src/main/java/com/google/api/codegen/discogapic/transformer/DiscoGapicNamer.java

```diff
@@ -119,30 +138,77 @@ public static Name getRequestName(Method method) {
     return Name.anyCamel(methodName, resourceName, "http", "request");
   }
 
+  /**
+   * Assuming the input is a child of a Method, returns the name of the field as a parameter. If the
+   * schema is a path or query parameter, then returns the schema's id(). If the schema is the
+   * request object, then returns "resource" appended to the schema's id().
+   */
+  public static Name getSchemaNameAsParameter(Schema schema) {
+    String paramString =
+        Strings.isNullOrEmpty(schema.reference()) ? schema.getIdentifier() : schema.reference();
+    String[] pieces = paramString.split("_");
+    Name param = Name.anyCamel(pieces);
+    if (Strings.isNullOrEmpty(schema.location())) {
+      param = param.join("resource");
+    }
+    return param;
+  }
+
   /** Get the request type name from a method. */
   public TypeName getRequestTypeName(Method method) {
     TypeNameConverter typeNameConverter = languageNamer.getTypeNameConverter();
     return typeNameConverter.getTypeNameInImplicitPackage(
         languageNamer.publicClassName(getRequestName(method)));
   }
 
-  /** Get the response type name from a method. */
+  public static Name getInterfaceName(String defaultInterfaceName) {
+    String[] pieces = defaultInterfaceName.split(REGEX_DELIMITER);
+    String resource = pieces[pieces.length - 1];
+    return Name.anyCamel(Inflector.singularize(resource), "admin");
+  }
+
+  /** Get the response type name from a method if the method has a non-null response type. */
+  @Nullable
   public static Name getResponseName(Method method) {
     if (method.response() != null) {
       String typeName =
           method.response().reference() != null
               ? method.response().reference()
               : method.response().getIdentifier();
       return Name.anyCamel(typeName);
-    } else {
-      String[] pieces = method.id().split(regexDelimiter);
-      String methodName = pieces[pieces.length - 1];
-      String resourceName = pieces[pieces.length - 2];
-      if (!method.isPluralMethod()) {
-        resourceName = Inflector.singularize(resourceName);
+    }
+    return null;
+  }
+
+  /**
+   * Get the canonical path for a method, in the form "(%s/\{%s\})+" e.g. for a method path
+   * "{project}/region/{region}/addresses", this returns "project/{project}/region/{region}".
```

Comment:
The literals should be plural in the example paths.

Reply:
Done.